### PR TITLE
Add tests that things throw when they should

### PR DESCRIPTION
**Pull request by segiddins**
\c @alazier @tgoyne

### CODE REVIEW

File: RealmSwift/Util.swift [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        return nil
+    }
+    return Int(index)
+}
+
+internal func throwException(message: String, userInfo: [String:AnyObject] = [:]) {
+    NSException(name: RLMExceptionName, reason: message, userInfo: userInfo).raise()
+}
+
+internal func assertPositive(int: Int, parameterName: String) {
```
Comment (alazier) [inline RealmSwift/Util.swift]: Might be better named `assertIntIsPositive`?

File: RealmSwift/Util.swift [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+// MARK: Internal Helpers
+
+internal func notFoundToNil(index: UInt) -> Int? {
+    if index == UInt(NSNotFound) {
+        return nil
+    }
+    return Int(index)
+}
+
+internal func throwException(message: String, userInfo: [String:AnyObject] = [:]) {
```
Comment (alazier) [inline RealmSwift/Util.swift]: `throwRealmException`?

File: RealmSwift/Util.swift (hunk 1 above)
Comment (segiddins) [inline RealmSwift/Util.swift]: No need, since we can always overload on that first parameter.

File: RealmSwift/Util.swift (hunk 2 above)
Comment (segiddins) [inline RealmSwift/Util.swift]: Inside Realm, I don't know what other kinds of exceptions we'd be throwing. I'll make the change if you really want it,

File: RealmSwift/Util.swift (hunk 1 above)
Comment (alazier) [inline RealmSwift/Util.swift]: Is there another case where we would actually overload this? When seeing this called in the code it is a bit confusing. I think being a bit more verbose here will improve readability.

File: RealmSwift/Util.swift (hunk 2 above)
Comment (alazier) [inline RealmSwift/Util.swift]: As is the case with `assertPositive` I just feel the name is a bit vague. It is not obvious what this does to someone unfamiliar with the codebase the first time they see this method. `throwRealmException` is unambiguously clear.

**Comment by alazier** [comment]
Other than naming ambiguity :+1:

File: RealmSwift/Util.swift (hunk 1 above)
Comment (tgoyne) [inline RealmSwift/Util.swift]: `assertPositive`'s also doubly-wrong: it's a precondition check and not an assertion, and it's checking for not negative numbers, not positive numbers.

File: Realm/RLMArrayLinkView.mm [hunk 3]
```diff
@@ -143,6 +143,9 @@ - (void)insertObject:(RLMObject *)object atIndex:(NSUInteger)index {
     RLMLinkViewArrayValidateInWriteTransaction(self);
 
     RLMValidateObjectClass(object, self.objectClassName);
+    if (index > _backingLinkView->size()) {
```
Comment (tgoyne) [inline Realm/RLMArrayLinkView.mm]: Should have an obj-c test for this.

File: RealmSwift/Util.swift (hunk 1 above)
Comment (segiddins) [inline RealmSwift/Util.swift]: @tgoyne in that case, what would you like it to be named?

File: Realm/RLMArrayLinkView.mm (hunk 3 above)
Comment (segiddins) [inline Realm/RLMArrayLinkView.mm]: Added

File: RealmSwift/Util.swift (hunk 2 above)
Comment (segiddins) [inline RealmSwift/Util.swift]: Done

**Comment by segiddins** [comment]
@tgoyne could you take a look at this?

**Comment by tgoyne** [comment]
Still uses the word `assert` for something that is not an assertion. `assertPositive(x)` should be a sanity check in a place where it is not *possible* for `x` to be non-positive, not a precondition check in a place where `x` is not allowed to be non-positive.

**Comment by segiddins** [comment]
@tgoyne changed the naming on `assertPositive`

**Comment by alazier** [comment]
I would prefer the names `ensurePositiveIndex(index: Int)` or `throwForNegativeIndex(index: Int)` and not take a second argument, since it is the same for all cases.

**Comment by jpsim** [comment]
@segiddins please just rename to `throwForNegativeIndex(index: Int)` and let's be done with this. You get my :+1:.

**Comment by segiddins** [comment]
Changed.

**Pull request merged**